zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java

```diff
@@ -88,13 +171,305 @@ private ConnectionDetails buildConnectionDetails(String zoneName, IdentityZone i
-            .create(uaaUriObject.toString().replace(currentUAAHostName, subdomain + "." + currentUAAHostName));
+            .create(uaaUriObject.toString().replace(currentUAAHostName, replacement));
         return newUAARoute.toString();
     }
 
     @Override
     public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
         this.publisher = applicationEventPublisher;
     }
+
+    private String getSubDomainStr(IdentityZone identityZone) {
+        String id = identityZone.getId();
+        String subDomain = identityZone.getSubdomain();
+        if(id.equals(subDomain)){
+            subDomain = null;
+        }
+        return subDomain;
+    }
+
+    public void createZone(OrchestratorZoneRequest zoneRequest) throws OrchestratorZoneServiceException,
+                                                                                   ZoneAlreadyExistsException,
+                                                                                   AccessDeniedException {
+        if (!IdentityZoneHolder.isUaa()) {
+            throw new AccessDeniedException("Zones can only be created by being authenticated in the default zone.");
+        }
+        String name = zoneRequest.getName();
+        String adminClientSecret = getAdminClientSecret(zoneRequest);
+
+        checkOrchestratorZoneExists(name);
+
+        String subdomain = zoneRequest.getParameters().getSubdomain();
+        String id = UUID.randomUUID().toString();
+        subdomain = getSubDomain(subdomain, id);
+
+        String zoneSigningKey = createSigningKey(subdomain);
+
+        IdentityZone identityZone = geIdentityZone(subdomain, name, id, zoneSigningKey);
+
+        IdentityZone created = createIdentityZone(identityZone);
+
+        if(created == null){
+            throw new OrchestratorZoneServiceException("Error while create identity zone for subdomain : " + subdomain );
+        }
+
+        createDefaultIdp(created);
+
+        createUserGroups(created);
+
+        createClient(adminClientSecret, subdomain, created);
+    }
+
+    private String getAdminClientSecret(OrchestratorZoneRequest zoneRequest) throws OrchestratorZoneServiceException {
+        String adminClientSecret = zoneRequest.getParameters().getAdminClientSecret();
+        if (!StringUtils.hasText(adminClientSecret)) {
+            throw new OrchestratorZoneServiceException(
+                "The \"" + UAA_ADMIN_SECRET + "\" field cannot contain spaces or cannot be blank.");
+        }
+        return adminClientSecret;
+    }
+
+    private String getSubDomain(String subdomain, String id) throws OrchestratorZoneServiceException {
+        String customSubdomain = getCustomSubdomain(subdomain);
+        if (customSubdomain == null) {
+            subdomain = id;
+        } else {
+            subdomain = customSubdomain;
+        }
+        return subdomain;
+    }
+
+    private void checkOrchestratorZoneExists(String name) throws ZoneAlreadyExistsException {
```

Comment:
This logic has to be handled with separate story created for this.

Reply:
Yes, But this case is handled here. may be there will separate story created for improvements.